### ISSUE TITLE
Awdb - active forecast point webservice

### ISSUE THREAD

**Issue by jdduh** [PSU-CSAR/awdb-retrieve#5]
We currently don't have a feature layer webservice for NWCC active forecasts. This layer will eventually replace the "stations USGS ACTIVE" service on NRCS AGOL and be used as the "master" AOI list (i.e., replacing Master AOI List Feature Service).

The current master AOI list is based on the stations USGS ACTIVE layer. The attribute denoting whether a USGS station is an active NWCC forecast location was provided by NWCC as a static layer. The attribute is not updated routinely.

The key attribute for identifying an active forecast is the "endDate". When the endData value is beyond the current date value (usually with a value of 2100-01-01 00:00:00) then the stations is an active forecast. The logic for creating a NWCC active forecasts is:

1. Use the active USGS stations to retrieve the stationtriplet values for each station
2. Use the [getStationMetadata](https://wcc.sc.egov.usda.gov/awdbWebService/webservice/testwebservice.jsf;jsessionid=s6XOYEzGDpCMUAnUcD66iKOHbaHjNq0waM7zGzU1.nrcsprd0383?webserviceName=/awdbWebService) method to retrieve the endDate values for each station
3. Keep the stations whose endDate is later then present date

For more information on AWDB webservices, see: https://www.nrcs.usda.gov/wps/portal/wcc/home/dataAccessHelp/webService/webServiceReference

Please refer to #4 for additional requirements.

**Comment by beautah** [comment]
@jdduh - this layer should NOT replace the active USGS layer, rather supplement it, I might suggest calling it "stations FCST ACTIVE" on agol - but you should really keep both since we store observed data for all the USGS network stations. You SHOULD base the AOI generation on the FCST stations though.

**Comment by beautah** [comment]
I'll add a step #4 as well:

4. Use the get [getForecastPoints](https://wcc.sc.egov.usda.gov/awdbWebService/webservice/testwebservice.jsf?webserviceName=/awdbWebService#) method to cross reference which ACTIVE USGS network stations are indeed FCST points - the "stationTriplet" is returned from this method and can be used to further filter the ACTIVE USGS STATIONS to ACTIVE FORECAST STATIONS

**Comment by lbross** [comment]
Just to clarify, does the getForecastPoints method return ALL forecast points or just ACTIVE forecast points? If just ACTIVE, I could skip checking the endDate values.

**Comment by lbross** [comment]
The BAGIS-Pro reporting tool also needs to query a winter start and end month at the basin level. We currently maintain this in the  PSU Master AOI List Feature Service. This new forecast point service will need to also query the PSU master service to get the most current values from these fields. BAGIS-Pro also needs the HUC2 value. Ensure that this is in the new webservice.

**Comment by jdduh** [comment]
NWCC suggests that use getForecastPoints (using wildcard) to get a list of all active forecast points, then use the getStationMetadata to get other forecast point attributes.

**Comment by jdduh** [comment]
The new stations_FCST_Active webservice will be disconnected from the USGS station database due to different stationTriplet ID values are used at NWCC. The basinarea data, which come from USGS, won't be available in the new webservice.

**Comment by lbross** [comment]
I've been experimenting with the [getForecastPoints()](https://wcc.sc.egov.usda.gov/awdbWebService/webservice/testwebservice.jsf?webserviceName=/awdbWebService#) api but can't find a combination of arguments to get ALL of the stationTriplets. Any suggestions?

**Comment by beautah** [comment]
@lbross - it appears the service is broken atm, you should be able to leave all the optional fields blank and get them all... but it appears that it only returns them when you give it stationIds. I'll look into the service, get that working again. Our new REST API can return forecast points and associated metadata, but you need to filter them

https://wcc.sc.egov.usda.gov/awdbRestApi/swagger-ui.html#/Station%20Metadata/getStations

this url pulls all the sites that could be a forecast point, if the object in the returned array has the "forecastPoint" key in it you know its a forecast point, so that's how you can filter. Here's the url to use for that (note that I'm pulling all USGS and BOR network sites as they are the networks that have forecast points): https://wcc.sc.egov.usda.gov/awdbRestApi/services/v1/stations?activeOnly=true&returnForecastPointMetadata=true&returnReservoirMetadata=false&returnStationElements=false&stationTriplets=*:*:USGS,*:*:BOR

in python you'd have something like:
```
import requests
url = "https://wcc.sc.egov.usda.gov/awdbRestApi/services/v1/stations?activeOnly=true&returnForecastPointMetadata=true&returnReservoirMetadata=false&returnStationElements=false&stationTriplets=*:*:USGS,*:*:BOR"
req = requests.get(url)
sites = req.json()
forecast_pnts = f = [i for i in sites if i.get("forecastPoint", False)]

```

I  would totally understand if you want to keep all this in the SOAP webservice, so I'll try to remember to let you know when that functionality is restored. Its 100% a bug.

**Comment by lbross** [comment]
Thanks @beautah. I'm just starting to familiarize myself with the current script to see what the best approach might be for populating this new service. I was unaware that some of the forecast stations could also be in the BOR network. My initial thought was to make a copy of the stations from the USGS_ACTIVE service and then to delete stations that aren't in the collection returned by getForecastPoints. This allowed me to re-use much of our existing code. But if these stations need to be merged with USGS_BOR, this might not be the best approach.

**Comment by beautah** [comment]
Yeah, if you hold out a little while you can just make a blank request to
the getForecastPoints method (the logicalAnd arg IS required however, the
rest can be ignored), that'll be the best way, no need to pass actual IDs.
It might take us a week or two to get the fix deployed however

**Comment by jdduh** [comment]
The PSU team has decided to use only the station information from AWDB based on the fact that the station ID values can be independently assigned in AWDB and cannot be correctly linked to USGS & BOR records. The current stations_USGS_ACTIVE webservice is derived from AWDB, but the PSU script also uses USGS api to get the basinarea data from USGS based on the station ID. This part of the script will be (has been?) deactivated.

If NWCC's active forecast station is always a subset of the stations_USGS_ACTIVE and that NWCC has resolved the station ID conflicts when pulling data from USGS & BOR, then Lesley's solution would be reasonable.

**Comment by lbross** [comment]
I'm still not clear if we need to merge stations_USGS_ACTIVE and stations_BOR_ACTIVE before checking for the awdb_id in the collection returned from getForecastPoints? Or if manipulating stations_USGS_ACTIVE can provide the full set of forecast stations? And it's no problem on our end to wait for the fix to be deployed.

**Comment by beautah** [comment]
There is no reason to use the triplets from USGS_ACTIVE or BOR_ACTIVE. Once
the SOAP method works again you can simply get the full list of forecast
triplets from the getForecastPoints by not using the filters at all (but
logicalAnd does need to be set to True). Once you have the list of forecast
triplets you can get use that to pull the metadata from USGS_ACTIVE and
BOR_ACTIVE to make a layer. @jdduh 's comment that USGS_ACTIVE represents
all the forecasts is not correct (there are a couple BOR as well), and
furthermore you cannot assume the station conflicts don't exist, they
absolutely do and will never not exist, ignore the fact these networks are
called BOR and USGS, they should really be called RES and GAGE or something
like that

so to summarize, you can get the triplets from getForecastPoints once we
get it fixed (a couple weeks I hope), then you can use that list of
triplets to make a FORECAST_ACTIVE layer by including the metadata from
USGS_ACTIVE and BOR_ACTIVE for that list of forecast triplets. You
basically will join a subset of BOR and USGS stations that have been
identified as forecast points.

**Comment by beautah** [comment]
Try the getForecastPoints method again, I discovered that it suddenly works
again without any fix on our end. Obviously there is an issue and we'll
look into it, but I checked last night and it was just magically working.

B

**Comment by lbross** [comment]
Glad that getForecastPoints is "fixed". Will a station ever be in both the BOR and USGS networks or are the two mutually exclusive?

**Comment by beautah** [comment]
since triplets are unique that's impossible, the network is in the triplet,
so you could have a USGS and BOR site in the same physical location but
they would be separate forecast points (one would forecast streamflow and
the other lake rise)

It appears that the getForecastPoints is not actually fixed too btw, we
have a pair of load balanced servers and it appears one of them is failing
while the other is not, so you have a 50/50 chance of it working. I'd hold
off coding against it for now.. although in the meantime you could simply
pass a list of all 50 states into the stateCd field and get a complete set,
that arg seems to be working.

**Comment by beautah** [comment]
the SOAP method is working again, it was a simple fix. You should be able to query the forecast points without including filters
![image](https://github.com/PSU-CSAR/awdb-retrieve/assets/6154031/09fe36d5-3c71-42bf-bd30-6e7f6d8ee849)

**Comment by lbross** [comment]
Reference syntax to set huc2 from forecast aoi master list:

1. reclass(!L0master_fcst_list_fs.huc2!)
2. def reclass(huc2):
    if (huc2 != None):
        return huc2
    else:
        return None

**Comment by lbross** [comment]
Prototype of new webservice [here](https://nrcs.maps.arcgis.com/home/item.html?id=79e8e6f9a175400988abee89093933e4). This webservice is not regularly updating yet and may be unavailable from time to time as it is worked on. I also set more strict permissions on it than the others to keep it out of general circulation.

**Comment by jdduh** [comment]
The stations_FCST_ACTIVE webservice looks good. Can you also transfer the bagis_note data from the master list to the new service? This attribute indicates the availability of forecast stations' associated AOI (GIS) data. I had to change the share setting for it to be used in BAGIS V3. Now it's back to its original share setting (NRCS organizational users only). Let me know if you want me to look at any specific aspect of the webservice.

**Comment by lbross** [comment]
It looks like bagis_note is a text field. How long should this field be? 100 characters? If the data in the FCST_ACTIVE webservice looks good, we can leave it publicly available. I had it restricted so that we weren't sharing bad data. At some point I'd like you to test changing data on the webservice to see if the changes persist when the webservice is updated. For example huc2 for an Alaska AOI or the bagis_note field. But I don't want to do this until the bagis_note field is implemented.

**Comment by jdduh** [comment]
Here are the unique values and their frequency counts (in parentheses) in the bagis_note field. The longest text string is 32 chars. 100 sounds good.

- "deprioritized by NWCC" (60 - mostly in CA)
- "excluded - has coincident site" (8)
- "excluded - poor data coverage" (3 - in AK, most AOI areas are outside US)
- "has AOI" (559)
- "too many sites - AOI too large" (17 - large AOIs that contain too many snow monitoring sites whose labels obscure the maps in the basin analysis report)

**Comment by lbross** [comment]
I have added the bagis_note field and republished the service. We are now ready to test editing the BAGIS-Pro fields in the service: huc2, winter_start_month, winter_end_month, and bagis_note. Can you try changing one or more of these fields on a couple of records via AGOL and then let me know? I will run the update script and see if it propogates the changes.
Do you know how the read/write security works with AGOL? I set the permissions to be the same as the other awdb webservices. I don't think we want non-admin users to be able to update this service. The other awdb services are reloaded during the weekly update.

**Comment by jdduh** [comment]
I have updated the huc2 and bagis_note values for the following forecast stations. They are new to the list. We can set the default huc2 value using the first two digits of the huc value that comes with awdb in the script. I will verify the AGOL read/write setting for the webservices.

huc, stationTri, name, huc2, bagis_note
160202040302, DELU1:UT:USGS, Dell Fk nr SLC, 16, New: colocate with 10171000:UT:USGS
160203100200, 10010000:UT:USGS, Great Salt Lake Inflow, 16, New: no AOI
170102080404, 00600:MT:USGS, Hellroaring Creek ab Reservoir nr Polson, 17, New: no AOI
180102030309, 11507002:OR:USGS, Upper Klamath Lake Local, 18, New: colocate with 11507001:OR:USGS
171100140403, 12097850:WA:USGS, White R bl Clearwater R nr Buckley, 17, New: no AOI
170102150701, 12393501:ID:USGS, Priest R Outflow NR Coolin, ID, 17, New: no AOI
130100021102, 08251500:CO:USGS, Rio Grande nr Lobatos, 13, New: no AOI (large size)
160203100200, 09UTGSLR:UT:BOR, Great Salt Lake Rise, 16, New: colocate with 10010000:UT:USGS
160501010500,  10337000:CA:BOR	Lake Tahoe, 16, New: colocate with (next to) 150017:CA:USGS

**Comment by lbross** [comment]
I think there is some confusion. I thought we were going to maintain the BAGIS values in the stations_FCST_ACTIVE service and eliminate the Master FCST List Feature Service. When I look at these AOIs in the stations_FCST_ACTIVE service, I don't see the updated values. The algorithm that updates the stations_FCST_ACTIVE service pulls the BAGIS values from the service before it reposts the updated stations.

I will look into updating the huc2 value from the huc value as a fallback for stations where it is null in the stations_FCST_ACTIVE service.

**Comment by jdduh** [comment]
Just to confirm the updates were made on the stations_FCST_ACTIVE feature layer, not on the master list. Please verify the values in AGOL or ArcGIS Hub.

**Comment by lbross** [comment]
I can see the updates now. I was/am accessing the service through ArcGIS Pro. I wonder if it takes some time for the updates to propagate initially?

**Comment by lbross** [comment]
This is done. Can you take another look at the service and see if things look good? The next steps would be to move the updated scripts to basins where they will automatically update early on Sunday mornings.

**Comment by jdduh** [comment]
I'm starting the list of BAGIS dependency of the stations FCST ACTIVE service. Please verify if the list is complete and confirm if the default settings are pointing to the stations FCST ACTIVE service.

1. BAGIS V3 source "Gauge Stations"
2. BAGIS Pro Batch Tool - Generate AOI Reports tool no longer needs to access the webservice
3. BAGIS Pro Batch Tool - Forecast Station Data tool

**Comment by lbross** [comment]
I will take a look at this next week. I would have the FCST ACTIVE service update for a few weeks with the task scheduler before we point the PC applications at it. I will deploy the FCST_ACTIVE service on basins next week too unless I hear otherwise.

**Comment by lbross** [comment]
The new service has been deployed on basins and the update script ran successfully. I would like to see it run for a couple of weeks before updating the clients. Action items on the client side:
- Update \vb-bagis\src\BAGIS_Reference_Maps.lyr to use new service as source for 'Gauge Stations' layer
- Update \bagis-pro\Settings\bagis.json to use new service url for gaugeStation parameter; Also used by bagis-pro
- Update \bagis-pro\Settings\bagis-pro.json to use new service url for MasterAoiList parameter. Note: We query the MasterAoiList when an AOI is set to get the current winter start/end month and huc2 values. We don't save these values locally. An AOI is set when a report is generated.

Note that BAGIS V3 users will have to use the 'Default Settings' button to retrieve these new settings when they are deployed.

Can I remove this warning from BAGIS V3? It shouldn't take too long and is confusing. It happens when you load the BAGIS settings form.
![image](https://github.com/PSU-CSAR/awdb-retrieve/assets/11527966/4a29cc10-2d96-426f-9e22-907b9093b53f)

**Comment by lbross** [comment]
Regarding the verbal request to generate a separate file when stations are added/removed. I could add a 'createdDate' to the service fairly easily. This would at least allow us to find the new ones easily...

**Comment by jdduh** [comment]
That's a good solution. With the same logic, I can check, for example, the bagis_note value for newly added stations. Let's ignore the request to keep track of the station updates in the script for now.

Just to confirm the removal of checking basinarea attribute in BAGIS V3.

**Comment by lbross** [comment]
So if bagis_note is null, it is a newly added station? That should work. There is only one station in the service currently where bagis_note is null: 06102500:MT:USGS. I will work on updating the BAGIS clients to use the new service.

**Comment by lbross** [comment]
I updated gaugeStation parameter in the bagis settings to point to the FCST ACTIVE service. This will affect BAGIS v3. I updated BAGIS Pro to use this parameter in lieu of the MasterAoiList parameter that we are sunsetting. I posted a new version of the BAGIS-Pro addIn to the basins FTP server. Please download and test. Once we confirm that this works, I will remove the MasterAoiList parameter from the parameter list.

FYI, the new service doesn't have the awdb_id field so this field will no longer be updated going forward. I still need to make the changes to BAGIS v3.

**Comment by lbross** [comment]
An update: I only needed to update the two .json configuration files on the server to repoint the gauge station layer to the new FCST ACTIVE service. The BAGIS_Reference_Maps.lyr didn't contain any references to the gauge stations so no changes were needed. I have removed the warnings about the missing basinarea field from BAGIS v3. To wrap this up I am waiting for @jduh to

1. Test the latest version of the BAGIS-Pro addIn from the basins FTP server. Note that the previous version might not work well because the field names are different in the new service
2. Update to the Default Settings in BAGIS v3. This should reset the gauge stations layer to FCST ACTIVE service. You will still get warnings about the basinarea field, but the AOI creation process should still work. It seemed to work for me. Once we know there are no additional changes required, I will publish a new version of this addIn.

**Comment by jdduh** [comment]
Confirm both features are working. Thanks!

**Comment by lbross** [comment]
Great! I just posted a new addIn to BAGIS v3 that eliminates the warning messages. And I have removed the MasterAoiList parameter from the configuration settings since BAGIS-Pro is now using the gaugeStation parameter which is pointed to the new FCST ACTIVE service.

**Comment by jdduh** [comment]
It seems that the stations_FCST_ACTIVE webservice hasn't been updated since May 2024. Also, the service is now shown as a geodatabase, instead of a feature layer. BAGIS Pro's Forecast Station Data tool seems to be working ok. It didn't give any error message the last time I used it.

**Comment by lbross** [comment]
The stations_FCST_ACTIVE webservice should be in the awdb folder. I'm not sure how it got deleted. I will look into this.

**Comment by jdduh** [comment]
I think several webservices got corrupted the last time I tried to publish the stations_FCST_ACTIVE_polygons_statistics_HU_16. ArcGIS Pro failed to publish the said service on NRCS AGOL. Afterwards, I started to notice some of the data contents do not match their names (for example, some polygon feature layers became point feature layers). I am going to delete the webservices that I know we don't need. Feel free to clean up NRCS AGOL if you see something that we don't need there. Hopefully this will tidy up the account.

**Comment by jdduh** [comment]
I have removed all the PRISM imageservices and several other files from the NRCS AGOL. Please don't delete the 4 items in the basin_analysis_reports folder for now.

**Comment by lbross** [comment]
AGOL won't let me share the feature service stations_FCST_ACTIVE. I tried renaming the service in the sd definition that still appeared in the awdb folder. That didn't work. I tried deleting the sd definition file and that still doesn't work. Can you think of any options besides opening a case with ESRI or publishing the service under a different name? BAGIS-Pro should be okay with a different name because it uses the item id in the parameter file. I will have to make some changes to the Python script that updates the service because it searches by name.

![image](https://github.com/user-attachments/assets/fe2c13de-9ccd-4a01-a01b-c386fb7376d0)

**Comment by jdduh** [comment]
We might need to delete stations_FCST_ACTIVE and republish it. I have a copy of the data that I downloaded on Oct 10, 2024 that I can use to republish the service. Let me know when I can try this approach.

**Comment by lbross** [comment]
Do you mean delete and re-add the stations_FCST_ACTIVE map from the ArcGIS Pro project? I can't find a way to delete the service from ArcGIS Online. You can try the approach any time. Let me know if you think it worked and I will try running the update script.

**Comment by jdduh** [comment]
stations_FCST_ACTIVE is republished. Everything looks normal. AGOL now has a recycle bin feature. We need to delete a webservice permanently to actually delete it. Otherwise, AGOL won't allow the publishing of the webservice with the same name even if the webservice is in the recycle bin. If your script uses the Delete function to remove a webservice, then you need to look for the permanent delete option/parameter to prevent issues.

**Comment by lbross** [comment]
So to clarify ... even though I renamed the sd definition and deleted it, you had to delete it out of the recycle bin before it was actually gone? The script does not use a Delete function. I don't think the script caused this problem. It's been running for months with no issues. I'll check now to make sure that it still works.

**Comment by jdduh** [comment]
AGOL recycle bin is a new feature. It was released in June 2024! What a coincidence!  https://www.esri.com/arcgis-blog/products/arcgis-online/announcements/recycle-bin-what-you-need-to-know/

API Changes
Changes to the existing APIs have been made to accommodate recycle bin. These include:

returning recycle bin support and status in portal calls
returning recycle bin items in item and content calls
allowing delete calls to bypass the recycle bin via the permanent delete call
Additionally, new endpoints allow restoring items and checking whether an item can be recycled or restored.

It is important to note that the default behavior of delete APIs has changed so that supported items are recycled rather than permanently deleted. Previous behavior can be replicated by passing a permanentDelete call as detailed in the doc pages below.

**Comment by lbross** [comment]
Any chance you could move it to the awdb folder with the other awdb services? This is where I look to check every 2 weeks to make sure the services updated? I also don't know if the script needs it there.

**Comment by jdduh** [comment]
Done!

**Comment by lbross** [comment]
I'm getting this error now when the Python script tries to publish the updated service. Any ideas?
https://support.esri.com/en-us/knowledge-base/error-user-cant-overwrite-this-service-using-this-data--000026287

**Comment by jdduh** [comment]
I will remove the firestats services. I think the SDs are all corrupted/crossed.  Here is the URL for the stations_FCST_ACTIVE, based on the ESRI page that you shared.

https://services.arcgis.com/SXbDpmb7xQkk44JV/arcgis/rest/services/stations_FCST_ACTIVE/FeatureServer

**Comment by lbross** [comment]
It looks like that fixed the problem. Here is an article from ESRI on their fuzzy search and how the api differs from the interactive search on AGOL. https://developers.arcgis.com/python/latest/guide/accessing-and-creating-content/. I tried to see if I could query by id, which would always be unique, but couldn't get it to work. 
If we really need to use that prefix for a feature layer service, it looks like there are options to examine the returned collection and be more precise. But for now things are good. Thanks for your help with this.
